[HOP] support generating schema for hop

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133645
* __->__ #133521

Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). Also extending the Basic types in c++ seems not that easy.

ydwu4 committed:
[HOP] support generate schema for hop

[ghstack-poisoned]

ydwu4 committed:
Update on "[HOP] support generate schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). 


[ghstack-poisoned]

ydwu4 committed:
Update on "[HOP] support generate schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). 


[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -102,9 +113,6 @@ def add(x: torch.Tensor, y: torch.Tensor):
     return pytree.tree_unflatten(result_flat, spec)
 
 
-associative_scan_op = HigherOrderOperator("associative_scan")
```
okay we should probably ban creating HigherOrderOp without subclassing it

ydwu4 replied:
Will do it in a separate diff.

zou3519 commented on test/export/test_export.py:
```diff
@@ -4160,6 +4169,11 @@ def false_fn(x):
+        schema = get_hop_schema(ep)
+        self.assertExpectedInline(
+            str(schema),
+            """cond(SymBool pred, GraphModule true_fn, GraphModule false_fn, Tensor[2] operands) -> Tensor[1] output""",
```
1) does export support Tensor[2] (instead of Tensor[]). If it does not, do we want to make export support this? (or just change it back to Tensor[])

ydwu4 replied:
Discussed with a few folks. The length can be used for better runtime input checking in Sigmoid. The conclusion is that we could keep it there and strip away the size if it's not required.

zou3519 commented on test/export/test_export.py:
```diff
@@ -5634,6 +5653,12 @@ def true_fn(x, y):
+        schema = get_hop_schema(exported_program)
+        self.assertExpectedInline(
+            str(schema),
+            """cond(Tensor pred, GraphModule true_fn, GraphModule false_fn, Tensor[3] operands) -> Tensor[1] output""",  # noqa: B950
```
please test something with more than 1 element in the tuple return

zou3519 commented on test/export/test_export.py:
```diff
@@ -179,6 +179,15 @@ def is_training_ir_test(test_name):
+        for node in ep.graph.nodes
+        if isinstance(node.target, torch._ops.HigherOrderOperator)
+    )
+    return torch._ops.HigherOrderOperator.generate_schema_from_fx_node(hop_node)
```
torch._library.utils.hop_schema(hop_node)

zou3519 replied:
(bikeshedding, staticmethod might be strange)

ydwu4 committed:
Update on "[HOP] support generating schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). Also extending the Basic types in c++ seems not that easy.


[ghstack-poisoned]

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -3545,6 +3545,144 @@ def cond_fn(x):
+                source=ConstantSource(
+                    f"__testing_hop_schema{len(shape_env.var_to_val)}"
+                ),
+                # dynamic_dim=DimDynamic.DYNAMIC,
+                # constraint_dim=None,
```
what's going on?

ydwu4 replied:
These shouldn't matter, will delete the commented part

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -3545,6 +3545,144 @@ def cond_fn(x):
             self.assertEqual(cnt.frame_count, 3)
 
 
+_hop_schema_test_dtypes = [
```
nit: this isn't a "dtype", call it something else

zou3519 replied:
Also I would match the casing in the schema anyways (SymInt instead of symint) to make it more consistent

zou3519 commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -37,6 +37,17 @@ def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
+    def __init__(self):
+        super().__init__("associative_scan")
+
+    def __call__(self, combine_fn, input, dim):
```
We should also probably put some docstrings here about waht each of the args are. One complaint for our aten ops is that they aren't well documented. Doesn't need to happen in this PR.

zou3519 commented on torch/_library/utils.py:
```diff
@@ -201,6 +201,44 @@ def zip_schema(
     return
 
 
+def hop_schema_from_fx_node(node):
+    from torchgen.gen_schema_utils import FunctionSchemaGen
+
+    hop = node.target
+    if not isinstance(hop, torch._ops.HigherOrderOperator):
+        raise RuntimeError("fx_node's target must be a hop.")
```
Does this work if the HOP takes a default value as input?

Also, if a HOP takes a default value as input, we shouldn't trace it into the graph (as per the rules for aten operators). I don't know if that's what happens today, we should file a follow-up issue for this if not

ydwu4 replied:
> if a HOP takes a default value as input, we shouldn't trace it into the graph

Each hop has its own logic nowadays. FlexAttention has a default kwarg `score_mod_other_buffers = tuple()` and it's currently traced into the graph. AssociativeScan is going to have a default kwarg reverse=False.

>  If someone adds a new default kwarg for a HOP

I'll write down my understanding about BC/FC: 

Yeah, this should work today. I can try to add a test for this but this should be after we actually use the schema gen mechanism for the control flow operators in export. 

**For BC:**
For example, associtive_scan adds `reverse=False`.The newer export receives an old hop without the default kwarg in its schema:
```
associative_scan(GraphModule combine_fn, Tensor[] input, int dim) -> Tensor[]
```
The new runtime is able to deserialize it by mapping the kwargs combined_fn, input and dim with proper inputs. For anything missing, (e.g. reverse=False), it will leave as it is. So we have BC in this case.
Case 2:  changing arg name name breaks BC:
For example, associative_scan changes `combine_fn` to `combine_graph` in its signature. The deserializer won't be able to map the `combine_fn` in old artifact to `combine_graph`. So this is BC-breaking.

**For FC:**
I think export generally doesn't care a lot about FC (at least that's my impression). Supposing a new artifact has `reverse=True` argument in associative_scan, when an old runtime tries to deserialize it, the `reverse=True` won't work if we create an associative_scan without the reverse argument. This means FC is broken if `reverse` is not the default value.

If `reverse=False`, (i.e. the BC-preserving change), the FC-story would be, we shouldn't add `reverse=False` into the graph. Then an older runtime can deserialize the op using the old one without any issue.

So concretely, i think we should:
- Do the`if a HOP takes a default value as input, we shouldn't trace it into the graph`
- Add a BC test when serde starts to use the generated schema.

zou3519 replied:
> So concretely, i think we should:

Yeah, that's how our current FC/BC things work for regular operators

zou3519 commented on torchgen/gen_schema_utils.py:
```diff
@@ -0,0 +1,90 @@
+    OperatorName,
+    Return,
+)
+
+
```
Put a docstring here that says something like "this isn't actually used in torchgen, it's some utilities from generating a schema from real arguments"

zou3519 reviewed (commented):
Some minor comments, please read. The default value one is important to test because of FC/BC.

ydwu4 committed:
Update on "[HOP] support generating schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). Also extending the Basic types in c++ seems not that easy.


[ghstack-poisoned]

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -3545,6 +3545,142 @@ def cond_fn(x):
+        )
+        self.assertExpectedInline(
+            str(schema3),
+            """test_op3(bool bool_v, int int_v, float float_v, str str_v, Tensor Tensor_v, SymInt SymInt_v, SymBool SymBool_v, GraphModule GraphModule_v, __torch__.torch.classes._Foo ScriptObj_v) -> (Tensor output, Tensor output)""",  # noqa: B950,
```
Uh, why do these both have the same output names

ydwu4 replied:
Oh this is because we pass in the same name of in ReturnGen. Thanks for pointing it ou!

The names of the output don't matter from a BC-FC point of view. We can just don't give them names (None) for simplicity. Updated the diff.

ydwu4 committed:
Update on "[HOP] support generating schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). Also extending the Basic types in c++ seems not that easy.


[ghstack-poisoned]

ydwu4 committed:
Update on "[HOP] support generating schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). Also extending the Basic types in c++ seems not that easy.


[ghstack-poisoned]

ydwu4 committed:
Update on "[HOP] support generating schema for hop"


Add a way of generating a FunctionSchema from example values because hop's schema varies even for the same hop.

We didn't use torch._C.FunctionSchema because we cannot construct the classes directly (e.g. "__init__" cannot be used for torch._C.FunctionSchema). Also extending the Basic types in c++ seems not that easy.


[ghstack-poisoned]
